ttnn/cpp/ttnn/types.hpp

```diff
@@ -59,76 +59,72 @@ struct CoreGrid {
-class Buffer : public tt::tt_metal::Buffer {
-    public:
         Buffer(Device *device, uint64_t size, uint64_t page_size, const BufferType buffer_type,
                 const TensorMemoryLayout buffer_layout = TensorMemoryLayout::INTERLEAVED,
                 std::optional< ShardSpecBuffer> shard_parameters = std::nullopt
             ) : tt::tt_metal::Buffer(device, size, page_size, buffer_type, buffer_layout, shard_parameters, false) {
-                this->buffer_id = GLOBAL_BUFFER_ADDRESS_MAP.get_buf_id(); // Each buffer has a unique ID
-                this->allocate();
+                // Private constructor used by factory method. Each buffer is given a unique id.
+                this->buffer_instance_id = Buffer::buf_id++;
             }
-        ~Buffer() {
-            this->deallocate();
-        }
-    private:
-        uint32_t buffer_id = 0;
+
         void allocate() {
             TT_ASSERT(this->device());
-            this->device()->push_work([this] () mutable {
-                bool bottom_up = this->buffer_type() == BufferType::DRAM;
-                tt::tt_metal::detail::AllocateBuffer(this, bottom_up);
+            this->device()->push_work([shared_buffer_ptr = shared_from_this()] () mutable {
+                bool bottom_up = shared_buffer_ptr->buffer_type() == BufferType::DRAM;
+                tt::tt_metal::detail::AllocateBuffer(shared_buffer_ptr.get(), bottom_up);
+                // Notify address getter that this buffer was allocated
+                shared_buffer_ptr->allocated_ = true;
+                shared_buffer_ptr->allocated_.notify_one();
                 // The address inserted here, will be used during asynchronous deallocate
-                GLOBAL_BUFFER_ADDRESS_MAP.insert(this->buffer_id, this->address());
-
+                Buffer::buf_id_to_address_map.insert({shared_buffer_ptr->buffer_instance_id, shared_buffer_ptr->address()});
             });
         }
+
         void deallocate() {
             if (this->device() == nullptr or not this->device()->initialized_ or this->size() == 0) {
                 return;
             }
             this->set_size(0);
             TT_ASSERT(this->device()->allocator_ != nullptr, "Expected allocator to be initialized!");
             // Extract the required buffer attributes from main thread (these are guaranteed to be correctly populated) and send to worker
-            this->device()->push_work([dev = this->device(), id = this->buffer_id, type = this->buffer_type()] () mutable {
-                // At this point, the address for this buffer has made it to GLOBAL_BUFFER_ADDRESS_MAP, since the worker has allocated the buffer.
-                tt::tt_metal::allocator::deallocate_buffer(*(dev->allocator_), GLOBAL_BUFFER_ADDRESS_MAP.buffer_address(id), type);
-                GLOBAL_BUFFER_ADDRESS_MAP.erase(id);
+            this->device()->push_work([dev = this->device(), id = this->buffer_instance_id, type = this->buffer_type()] () mutable {
```

Comment:
I think you can wait address atomic in desctructor. It will almost never happen, in this case. But you need global map for only one reason.
Also there is a chance that AsyncBuffer destructor will be called in the device thread after allocation. Can we call push_work from the device thread/

Reply:
`I think you can wait address atomic in desctructor. It will almost never happen, in this case. But you need global map for only one reason.`
Waiting for the buffer to be allocated in the destructor will stall the main thread. For cases where deallocations are frequent during the inference/training loop, this is would be sub-optimal, since the worker threads would be flushed each time. This can (and has happened) in workloads where the main thread is significantly ahead of workers and tries to deallocate a buffer that is currently not allocated.

`Also there is a chance that AsyncBuffer destructor will be called in the device thread after allocation. Can we call push_work from the device thread/`

Yes, we can safely call `push_work` from the worker thread. It will be processes as a JIT call.

Reply:
I think if we just can make address to be shared_ptr<atomic_uint> address_ptr; and just pass it to the job.